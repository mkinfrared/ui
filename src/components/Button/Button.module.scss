@import "../../styles/base/baseButton";

%button {
  @extend %button-base;

  min-width: 10rem;
  padding: 0.5rem 2rem;

  &.loading {
    color: transparent;

    > * {
      opacity: 0;
      transform: scale(0);
    }

    .spinner {
      opacity: 1;
      transform: translate(-50%, -50%) scale(1);
    }
  }

  svg {
    fill: currentcolor;
    height: var(--text-size);
  }

  .content {
    align-items: center;
    display: inline-flex;
    gap: 1rem;
    max-height: var(--text-height);
    white-space: nowrap;
  }

  > .spinner {
    fill: var(--spinner-color);
    height: var(--text-height);
    left: 50%;
    opacity: 0;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%) scale(0);
  }
}

.Button {
  @extend %button;
}
