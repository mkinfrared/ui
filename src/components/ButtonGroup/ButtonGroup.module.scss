@use "../../styles/base/baseButton" as bb;

.ButtonGroup {
  @include bb.button-border-radius;

  --button-border-color: var(--button-group-border-color, var(--white));
  --button-background: transparent;

  background: var(--night-call);
  border-radius: var(--border-radius);
  display: inline-flex;
  flex-wrap: nowrap;
  width: fit-content;

  &.secondary {
    background: var(--secondary-gradient);
  }

  &.success {
    background: var(--success-gradient);
  }

  &.error {
    background: var(--error-gradient);
  }

  &.outlined {
    --button-border-color: var(--primary);

    background: transparent;
    border: 2px solid var(--button-border-color);
    color: var(--primary);

    &.secondary {
      --button-border-color: var(--secondary);

      color: var(--secondary);
    }

    &.success {
      --button-border-color: var(--success);

      color: var(--success);
    }

    &.error {
      --button-border-color: var(--error);

      color: var(--error);
    }

    .button {
      color: inherit;

      &:first-child {
        padding-left: calc(2rem - 2px);
      }

      &:last-child {
        padding-right: calc(2rem - 2px);
      }
    }
  }

  .button {
    background: transparent;
    border-radius: 0;
    margin: 0;
    min-width: 4rem;

    &::before {
      background-color: #ffffff;
      content: "";
      inset: 0;
      opacity: 0;
      position: absolute;
      transition: opacity 500ms ease-in;
    }

    + .button {
      border-left: 1px solid var(--border-color);

      &:not(:last-child) {
        border-radius: 0;
      }
    }

    &:first-child {
      border-radius: var(--border-radius) 0 0 var(--border-radius);
    }

    &:last-child {
      border-radius: 0 var(--border-radius) var(--border-radius) 0;
    }

    &:active {
      transform: scale(1);

      > span {
        filter: brightness(0.8);
        transform: scale(0.97);
      }
    }

    &:hover {
      &::before {
        opacity: 0.1;
      }
    }
  }
}
