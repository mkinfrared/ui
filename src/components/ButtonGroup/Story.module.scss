.Story {
  .container {
    margin-bottom: 2rem;
  }

  .buttonContainer {
    margin-bottom: 1rem;
  }
}
