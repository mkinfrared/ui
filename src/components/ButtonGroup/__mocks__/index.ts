export { default } from "./ButtonGroup.mock";
