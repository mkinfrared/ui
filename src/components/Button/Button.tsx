/* eslint-disable react/button-has-type */

import Spinner from "components/Spinner";
import { classNames } from "utils";

import css from "./Button.module.scss";
import { ButtonColors, ButtonProps, ButtonVariants } from "./Button.type";

/*
  renders an html button

  @example
  see Button.stories.tsx
 */

const colors: Record<ButtonColors, string> = {
  primary: css.primary,
  secondary: css.secondary,
  success: css.success,
  error: css.error,
};

const variants: Record<ButtonVariants, string> = {
  contained: css.contained,
  outlined: css.outlined,
};

const getButtonColor = (color: ButtonColors) => colors[color];
const getButtonVariant = (variant: ButtonVariants) => variants[variant];

const Button = ({
  buttonRef,
  children,
  className,
  color = "primary",
  disabled = false,
  loading = false,
  type = "button",
  variant = "contained",
  ...rest
}: ButtonProps) => {
  const classes = [
    css.Button,
    className,
    getButtonVariant(variant),
    getButtonColor(color),
  ];

  if (loading) {
    classes.push(css.loading);
  }

  return (
    <button
      className={classNames(...classes)}
      disabled={disabled}
      data-testid="Button"
      ref={buttonRef}
      type={type}
      {...rest}
    >
      <Spinner className={css.spinner} type="bars" />
      <span className={css.content}>{children}</span>
    </button>
  );
};

export default Button;
