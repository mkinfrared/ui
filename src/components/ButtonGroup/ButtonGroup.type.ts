import { ReactNode } from "react";

import { ButtonProps } from "components/Button";

export type ButtonGroupProps = Pick<ButtonProps, "color" | "variant"> & {
  children?: ReactNode;
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
};
