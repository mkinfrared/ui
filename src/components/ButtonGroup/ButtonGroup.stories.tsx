import { action } from "@storybook/addon-actions";
import { Meta, Story } from "@storybook/react";
import React from "react";

import Button from "components/Button";
import Heading from "components/Heading";
import IconButton from "components/IconButton";
import { ReactComponent as Doc } from "icons/description.svg";
import { ReactComponent as Eye } from "icons/visibility.svg";

import { ButtonGroup } from "./ButtonGroup";
import { ButtonGroupProps } from "./ButtonGroup.type";
import css from "./Story.module.scss";

export default {
  title: "UI/ButtonGroup",
  component: ButtonGroup,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const Template: Story<ButtonGroupProps> = (args) => (
  <div className={css.Story}>
    <div className={css.container}>
      <Heading>Default</Heading>
      <ButtonGroup {...args}>
        <Button onClick={action("click")}>Lorem</Button>
        <Button onClick={action("click")}>doloribus</Button>
        <Button onClick={action("click")}>elit</Button>
        <Button onClick={action("click")}>consectetur</Button>
      </ButtonGroup>
    </div>
    <div className={css.container}>
      <Heading>Variants</Heading>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="contained">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="outlined">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
    </div>
    <div className={css.container}>
      <Heading>Colors</Heading>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="contained" color="primary">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="contained" color="secondary">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="contained" color="success">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="contained" color="error">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="outlined" color="primary">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="outlined" color="secondary">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="outlined" color="success">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
      <div className={css.buttonContainer}>
        <ButtonGroup {...args} variant="outlined" color="error">
          <Button onClick={action("click")}>Lorem</Button>
          <Button onClick={action("click")}>doloribus</Button>
          <Button onClick={action("click")}>elit</Button>
          <Button onClick={action("click")}>consectetur</Button>
        </ButtonGroup>
      </div>
    </div>
  </div>
);

const WithIconTemplate: Story<ButtonGroupProps> = () => (
  <div className={css.Story}>
    <div className={css.container}>
      <ButtonGroup>
        <IconButton onClick={action("click")}>
          <Doc />
        </IconButton>
        <Button onClick={action("click")}>consectetur</Button>
      </ButtonGroup>
    </div>
    <div className={css.container}>
      <ButtonGroup>
        <Button onClick={action("click")}>Lorem ipsum</Button>
        <IconButton onClick={action("click")}>
          <Eye />
        </IconButton>
      </ButtonGroup>
    </div>
  </div>
);

const Default = Template.bind({});
const WithIcon = WithIconTemplate.bind({});

Default.args = {
  color: "primary",
};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default, WithIcon };
