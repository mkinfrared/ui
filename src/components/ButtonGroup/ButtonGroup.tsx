import {
  Children,
  ReactElement,
  ReactNode,
  cloneElement,
  isValidElement,
} from "react";

import { ButtonColors, ButtonProps, ButtonVariants } from "components/Button";
import { classNames } from "utils";

import css from "./ButtonGroup.module.scss";
import { ButtonGroupProps } from "./ButtonGroup.type";

const colors: Record<ButtonColors, string> = {
  primary: css.primary,
  secondary: css.secondary,
  success: css.success,
  error: css.error,
};

const variants: Record<ButtonVariants, string> = {
  contained: css.contained,
  outlined: css.outlined,
};

const getButtonColor = (color: ButtonColors) => colors[color];
const getButtonVariant = (variant: ButtonVariants) => variants[variant];

const ButtonGroup = ({
  className,
  children,
  color = "primary",
  variant = "contained",
}: ButtonGroupProps) => {
  const classes = [
    css.ButtonGroup,
    className,
    getButtonVariant(variant),
    getButtonColor(color),
  ];

  const childrenWithProps = Children.map<ReactNode, ReactNode>(
    children,
    (child) => {
      if (!isValidElement(child)) {
        return child;
      }

      const buttonClassNames = classNames(child.props.className, css.button);

      const clone = cloneElement<ButtonProps>(
        child as ReactElement<ButtonProps>,
        {
          className: buttonClassNames,
          variant: "contained",
          color: "primary",
        },
      );

      return clone;
    },
  );

  return (
    <div className={classNames(...classes)} data-testid="ButtonGroup">
      {childrenWithProps}
    </div>
  );
};

export default ButtonGroup;
