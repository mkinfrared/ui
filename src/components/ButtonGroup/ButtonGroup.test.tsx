import { render } from "@testing-library/react";
import React from "react";

import Button from "components/Button";

import { ButtonGroup } from "./ButtonGroup";

jest.mock("components/Button");

describe("<ButtonGroup />", () => {
  const Component = (
    <ButtonGroup>
      <Button />
      <Button />
      <Button />
    </ButtonGroup>
  );

  it("should be defined", () => {
    expect(ButtonGroup).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("ButtonGroup");

    expect(element).toBeDefined();
  });
});
