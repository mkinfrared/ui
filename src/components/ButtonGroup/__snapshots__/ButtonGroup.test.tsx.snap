// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> should match the snapshot 1`] = `
<div>
  <div
    class="ButtonGroup contained primary"
    data-testid="ButtonGroup"
  >
    <div
      data-testid="Button"
    >
      {"className":"button","variant":"contained","color":"primary"}
    </div>
    <div
      data-testid="Button"
    >
      {"className":"button","variant":"contained","color":"primary"}
    </div>
    <div
      data-testid="Button"
    >
      {"className":"button","variant":"contained","color":"primary"}
    </div>
  </div>
</div>
`;
